hal/src/rtl872x/exflash_hal.cpp

```diff
@@ -291,36 +595,61 @@ static bool is_block_erased(uintptr_t addr, size_t size) {
 }
 
 __attribute__((section(".ram.text"), noinline))
-static int erase_common(uintptr_t start_addr, size_t num_blocks, int len) {
-    ExFlashLock lk;
-
-    const size_t block_length = len == EraseSector ? 4096 : 64 * 1024;
-
-    /* Round down to the nearest block */
-    start_addr = ((start_addr / block_length) * block_length);
-
-    for (size_t i = 0; i < num_blocks; i++) {
-        if (!is_block_erased(start_addr + i * block_length, block_length)) {
-            FLASH_Erase(len, start_addr + i * block_length);
-        }
+static int erase_common(uintptr_t address, hal_exflash_cmd_t cmd) {
+    ExFlashLock lk(false);
+    size_t size = 0;
+    if (cmd == GENERIC_FLASH_CMD_SE) {
+        size = 4 * 1024;
+    } else if (cmd == GENERIC_FLASH_CMD_BE32K) {
+        size = 32 * 1024;
+    } else if (cmd == GENERIC_FLASH_CMD_BE64K) {
+        size = 64 * 1024;
+    } else {
+        return SYSTEM_ERROR_INVALID_ARGUMENT;
     }
+    address = ((address / (size)) * (size));
+    if (is_block_erased(address, (size))) {
+        return size;
+    }
+    RtkSpicFlashConfigGuard guard;
 
-    // LOG_DEBUG(ERROR, "Erased %lu %lukB blocks starting from %" PRIxPTR,
-    //           num_blocks, block_length / 1024, start_addr);
-
-    dcacheCleanInvalidateAligned(SPI_FLASH_BASE + start_addr, block_length * num_blocks);
-
-    return SYSTEM_ERROR_NONE;
+    rtkSpiFlashSendCmd(GENERIC_FLASH_CMD_WREN);
+    /* Disable SPI_FLASH */
+    SPIC->ssienr = 0;
+    rtkSpiSetTxMode();
+    /* set flash cmd + addr and write to fifo */
+    SPIC->dr[0].word = RTK_SPIC_CMD_ADDR_FORMAT(cmd, address);
+    rtkSpiFlashEnableCsWriteErase();
+    rtkSpiFlashSendCmd(GENERIC_FLASH_CMD_WRDI);
+    return size;
 }
 
 __attribute__((section(".ram.text"), noinline))
 int hal_exflash_erase_sector(uintptr_t start_addr, size_t num_sectors) {
-    return erase_common(start_addr, num_sectors, EraseSector);
+    size_t sizeInBytes = num_sectors * 4096;
+    uintptr_t addr = start_addr;
+    while (sizeInBytes > 0) {
+        hal_exflash_cmd_t cmd = GENERIC_FLASH_CMD_SE;
+        if (addr % (64*1024) == 0 && sizeInBytes >= 64*1024) {
```

Comment:
This checks if we are being asked to erase a 64k or 32k aligned series of sectors, and if so, coalesces them into a more efficient Block Erase command, right? 
Is there a concern that `start_addr` might not be sector aligned in the single sector erase case?

Reply:
Yes, exactly. No concern as you are supposed to give a sector address, if it happens to be an address within the sector not at the start, it should still work.

Reply:
Yes, I saw in the data sheet the sector erase command will erase the entire sector of the given address, even if it is not exactly at the start. 
I can see `hal_storage_erase()` calls `hal_exflash_erase_sector()`. Is there a concern that a user would call `hal_storage_erase()` with a non-sector aligned address and arbitrary size, causing a sector erase to erase data before the address desired?

Reply:
> causing a sector erase to erase data before the address desired

That is the intended behavior even in this case. And more importantly storage HAL is an unstable/experimental API, not for general use.